### DISCUSSION TITLE
[8일차-지원] Css-in-Js vs Css-in-Css의 차이

### DISCUSSION THREAD

**Discussion by smarfy99** [zzangCS/zzangInterview#30] [프론트 면접 질문]
각각의 특징과 차이를 말해보세요.

**Comment by ez1n** [comment]
- **CSS의 문제점**

** 동적인 기능 요구가 증가하고 점점 복잡해지면서 HTML과 CSS만으로 화면의 모든 스타일을 제어할 수 없는 상황이 되었다.

1. 모든 스타일이 전역에 선언되어 중복되지 않는 클래스 이름을 적용해야함
2. CSS간의 의존관계
3. 유지보수 문제
4. JS코드와 상태값을 공유할 수 없는 문제

이러한 CSS의 문제점을 해결하기 위해 CSS-in-CSS방식과 CSS-in-JS 방식이 등장함

- **CSS-in-CSS (CSS 모듈화, CSS 전처리기)**

순수 CSS와 마찬가지로 전체 페이지에 필요한 CSS를 로딩하여 style 태그를 생성한다.

**CSS 모듈** : CSS를 모듈화하여 사용하는 방식으로 CSS 클래스를 만들면 자동으로 고유한 클래스 이름을 만들어 스코프를 지역적으로 제한한다.

**전처리기** : CSS와 유사하지만 중첩, 조건문, 반복문 등의 기능을 제공한다. (자신만의 특별한 구문을 가지고 CSS를 생성)
⇒ CSS의 문제점을 일반적인 프로그래밍 개념을 사용하여 보완함 (Sass, Less 등)


- **CSS-in-JS**
자바스크립트 코드에서 CSS를 작성하는 방식

⇒ 자바스크립트를 사용하여 스타일을 선언적이고 유지보수 가능한 방식으로 설명한다.
⇒ JS를 CSS로 전환하는 고성능 컴파일러로 런타임 및 서버 사이드에서 작동
 ⇒ CSS를 컴포넌트 레벨로 추상화할 수 있다.
ex) Styled Components

- **차이점**
1. css-in-js는 동적 스타일 적용에 자유롭고 특정 컴포넌트에 맞는 스타일이 종속되며 컴포넌트가 렌더링 될 때만 스타일 태그를 생성하여 컴포넌트 기반의 프레임워크를 사용할 때 용이하다.
2. css-in-css는 클래스 이름의 중복 문제가 해결되고 네이밍 규칙을 간소화할 수 있다. 렌더링할 때 모든 CSS 스타일 요소를 로딩하기 때문에 인터렉티브한 웹 사이트에서 바로 적용이 가능하여 사용자 편의성에서 뛰어나다.

**Reply by ez1n** [reply]
https://s-core.co.kr/insight/view/%EC%9B%B9-%EC%BB%B4%ED%8F%AC%EB%84%8C%ED%8A%B8-%EC%8A%A4%ED%83%80%EC%9D%BC%EB%A7%81-%EA%B4%80%EB%A6%AC-css-in-js-vs-css-in-css/
